nemo/collections/nlp/models/language_modeling/lm_model.py

```diff
@@ -157,47 +170,47 @@ def validation_step(self, batch, batch_idx):
         nsp_loss = self.nsp_loss(logits=nsp_logits, labels=labels)
 
         loss = self.agg_loss(loss_1=mlm_loss, loss_2=nsp_loss)
-
         perplexity = self.perplexity_metric(mlm_loss)
         tensorboard_logs = {'val_loss': loss, 'perplexity': perplexity}
         return {'val_loss': loss, 'log': tensorboard_logs}
 
     def validation_epoch_end(self, outputs):
+        """Called at the end of validation to aggregate outputs.
+
+        Args:
+            outputs (list): The individual outputs of each validation step.
+
+        Returns:
+            dict: Validation loss and tensorboard logs.
         """
-        Called at the end of validation to aggregate outputs.
-        :param outputs: list of individual outputs of each validation step.
-        """
-        avg_loss = torch.stack([x['val_loss'] for x in outputs]).mean()
-        perplexity = torch.stack([x['log']['perplexity'] for x in outputs]).mean()
-        tensorboard_logs = {'val_loss': avg_loss, 'perplexity': perplexity}
-        logging.info(f"evaluation perplexity {perplexity.item()}")
-        return {'val_loss': avg_loss, 'log': tensorboard_logs}
-
-    def setup_training_data(self, train_data_layer_config: Optional[Dict]):
-        if 'shuffle' not in train_data_layer_config:
-            train_data_layer_config['shuffle'] = True
+        if outputs:
+            avg_loss = torch.stack([x['val_loss'] for x in outputs]).mean()
+            perplexity = torch.stack([x['log']['perplexity'] for x in outputs]).mean()
+            tensorboard_logs = {'val_loss': avg_loss, 'perplexity': perplexity}
+            logging.info(f"evaluation perplexity {perplexity.item()}")
+            return {'val_loss': avg_loss, 'log': tensorboard_logs}
+
+    def setup_training_data(self, train_data_config: Optional[DictConfig]):
         self._train_dl = (
-            self.__setup_preprocessed_dataloader(train_data_layer_config)
+            self._setup_preprocessed_dataloader(train_data_config)
             if self.tokenizer is None
-            else self.__setup_text_dataloader(train_data_layer_config)
+            else self._setup_dataloader(train_data_config)
         )
 
-    def setup_validation_data(self, val_data_layer_config: Optional[Dict]):
-        if 'shuffle' not in val_data_layer_config:
-            val_data_layer_config['shuffle'] = False
+    def setup_validation_data(self, val_data_config: Optional[DictConfig]):
         self._validation_dl = (
-            self.__setup_preprocessed_dataloader(val_data_layer_config)
+            self._setup_preprocessed_dataloader(val_data_config)
```

Comment:
This logic is duplicated in both setup_training_data and setup_validation_data. How about merging _setup_preprocessed_dataloader and _setup_dataloader into one method? Is it better?

Reply:
I would rather leave that to @yzhang123 to decide/implement. I don't think we need to change that in this PR.